Add options to Mail Account settings to improve inbound mail processing

ivan-compucorp opened the pull request:
Overview
----------------------------------------
According to new requirements we need to add an additional options to the **Mail Account** settings (see _Mail Accounts_ page `/civicrm/admin/mailSettings?action=add&reset=1`) to change the email to activity processing behavior - when `Email-to-Activity Processing` value is selected for **Used For?** field we need to:
1. Show an additional checkbox field - **Skip emails which do not have a Case ID or Case token** (defaults to unticked).
If checked: emails which do not have a valid case ID or case token should be moved into the `ignored` mail folder after processing and no activity should be created.
1. Show an additional checkbox field - **Do not create new contacts when filing emails** (defaults to unticked).
If checked: when CiviCRM checks for a matching contact, if no matching contact is found it will not create one and the email is filed.
So now administrator should be able to configure whether CiviCRM will create a new contact where a contact does not already exist within the system when filing inbound emails, so that he can ensure that new contacts are not created in the system when filing an email. The email should still be filed, even if no contacts are matched.

Email is considered to be related to cases if it's subject matches any of the following conditions:
- Starts with `[<any string> #<any string>]`, e.g: `[case #2cd8467]`, `[policy initiative #bdf3896]`.
- Ends with `Case ID <number>`, e.g: `Case ID 1298`, `case id 9`.

Also:
- Label of new field is printed to the right of the checkbox, not to the left, because left side is thin and long labels would look weird there.
- _Activity Status_ field label is also moved to the right for consistency. Thanks to this approach these fields would look more like _Used For?_ dependent fields, and this would make them look more intuitive hence will improve UX.


Before
----------------------------------------
Without the new options above enabled CiviCRM will process emails as normal i.e:
1. Will file emails without a case ID or case hash in the subject line against a contact record.
1. Will always create new contacts if no matching contact found.

And _Activity Status_ field label location is not changed.
![image](https://user-images.githubusercontent.com/39520000/94404970-82608900-0178-11eb-8787-7ccd26227cf8.png)

After
----------------------------------------
With the new options above enabled CiviCRM:
1. Will skip any emails that do not have either the case ID or case hash in the subject line and will instead move those emails to the `ignored` folder.
1. Will not create new contacts if no matching contact found (from / to / cc / bcc), but email still will be filed.

Now form looks like this:
![image](https://user-images.githubusercontent.com/39520000/94405049-a8862900-0178-11eb-9fc5-08cc43c6fdd7.png)

Technical Details
----------------------------------------
To make the **Skip emails which do not have a Case ID or Case token** checkbox work we:
1. Added new field to the `civicrm_mail_settings` db table - `is_non_case_email_skipped`. Respective upgrader is added.
1. Updated schema `MailSettings.xml` file to have new field.
1. Updated form related files - `Form/MailSettings.php` and `MailSettings.tpl` - to print new field and save the value on form submit. As it has long help text it was desided to move it to `MailSettings.hlp` file and use civicrm ajax help popup feature.
1. Updated `EmailProcessor.php` to handle new field during inbound email processing.
1. The `CRM_Utils_Mail_CaseMail` class is added to store `isCaseEmail()` method, which is used to check if email is related to cases or not before email processing (creating new activity). The regexp patterns are copied from `CRM_Activity_BAO_Activity->create()` method for consistency, as it also performs this check.
1. The `caseEmailSubjectPatterns()` hook is added to make it possible to change default email subject regexp patterns. It may be useful for example for _CiviCase_ extension to add case type category names to patterns.
1. Unit tests were added to test processing (with `is_non_case_email_skipped` enabled) of 2 emails: with case hash in subject and without case hash.
1. Also as part of this PR formatting of `MailSettings.tpl` file is fixed.

To make the **Do not create new contacts when filing emails** checkobx work we:
1. Added new field to the `civicrm_mail_settings` db table - `is_contact_creation_disabled_if_no_match`. Respective upgrader is added.
1. Updated `DAO/MailSettings.php` and schema `MailSettings.xml` files to have new field.
1. Updated form related files - `Form/MailSettings.php` and `MailSettings.tpl` - to print new field and save the value on form submit.
1. Updated `EmailProcessor.php` to handle new field during inbound email processing. This also involved updating the `Incoming.php` to make contact creation optional.
1. Unit tests were added to test processing with `is_contact_creation_disabled_if_no_match` enabled.

### About the new `CRM_Utils_Mail_CaseMail` class
This class has a potential to become generic and widely used class, thats why:
• Email subject patterns are not just hardcoded, but are reusable, see `getSubjectPatterns()` method.
• The `getSubjectPatterns()` method is written keeping in mind `CiviCase` extension (thats why  `caseEmailSubjectPatterns()` hook is added), which has case type categories, which names could be used in email subject, e.g: `[case #hahoheh] Super email` -> `[policy initiative #hahoheh] Super email`.

For now the most useful part of this class is `isCaseEmail()` method, but some other methods may be added later. For example:
1. Generate case hash - there is a code duplication in civicrm to generate case hash, i.e:
`$hash = substr(sha1(CIVICRM_SITE_KEY . $form->_caseId), 0, 7)`.
1. Generate case subject - there is a code duplication in civicrm to generate case related mail subject, i.e:
`$subject = "[case #$hash] $subject"`.

⚠  The email regexp patterns are copied from `CRM_Activity_BAO_Activity->create()` and now it makes sense to refactor that method to use `CRM_Utils_Mail_CaseMail->getSubjectPatterns()` to get patterns. But will skip it now as it's not in the scope of the task.

### One more comment
I've noticed that civicrm stops processing of inbound email if it failed to connect to some mail account. This may cause some issues in case if there are several accounts created: if processing would be stopped on first account - then no other accounts would be processed as well.
I guess this should be added to the todo list to fix in the nearest future.

mattwire commented:
@ivan-compucorp Can you squash your commits?

I'm not sure about adding two more columns for settings, maybe a serialised field would be better as these settings are specific to a use-case (ie. email-to-activity) and not to all inbound email settings. @colemanw thoughts?

ivan-compucorp commented:
@mattwire do you want to squash all commits to one? Or to two maybe? Why should we merge it btw?
Also is it okay to use `5.30.0.mysql.tpl` file name for db upgrader?

mattwire commented:
@ivan-compucorp It is helpful to have one commit for each bit, so if it makes sense to squash into 2-3 that's fine. Basically the 19 commits would just clutter up the core git history and we don't need to see how it was developed, only that it was developed. It is also easier to merge/revert a single commit if there are issues.

Leave the upgrader for now - this will need reviewing thoroughly (by someone else) and we'll know which version it will hit then.

eileenmcnaughton commented:
@ivan-compucorp @mattwire so there is no hard & fast rule but basically each commit should standalone - ie I should be able to pull off the first commit into a separate PR & it should be OK to merge without any others (although with schema updates that can just mean the schema update is complete in itself). You might ease this through review by picking out some small ones like https://github.com/civicrm/civicrm-core/pull/18624/commits/6fceec4de7c6aac838a370f412b83022101b1235  and https://github.com/civicrm/civicrm-core/pull/18624/commits/6fceec4de7c6aac838a370f412b83022101b1235 into separate PRs that can be merged fairly quickly, reducing the review load

However, I think @demeritcowboy is up for reviewing this & if so you might want to see what will work for him.

demeritcowboy commented:
I'm up for reviewing it as one big thing and then squashing or separating where appropriate before final merge. But yes I'd agree something like whitespace fixes like that one can be pulled out into another PR.

ivan-compucorp commented:
I see. I don't think it worth to create a separate PR for this, but it could be a separate commit.
I'll wait for comments (if any) and squash commits after that, okay?

ivan-compucorp commented:
@demeritcowboy I know that you may be busy, at the same time I let me point that one of our clients wants this change on his site next month. Afair civicrm would be released on Oct 7 (next week), so if we can merge this PR and have it in this release - that would be great.
Let me know if there is something I can do to make it happen.

demeritcowboy commented:
Hi I'm planning to look this week. But FYI master (5.31) where new features get added won't be released until early November. Next week's release is 5.30 which right now is just taking important bug fixes.

And I suppose if it's just one site you could apply this as a patch there in the meantime?

eileenmcnaughton commented:
@ivan-compucorp the reason we split off easy commits sometimes is that for most PRs one of the biggest tasks is getting through through review so anything that makes the reviewer's job easier helps with that - reducing the amount of code they have to parse in one sitting can be a big part of that. However, you should take your queues from @demeritcowboy who has agreed to review this (which is a non-trivial thing to do)

ivan-compucorp commented:
@demeritcowboy we can't apply this as a patch, because it contains db structure changes.

P.S. Just want to add FYI that this code was tested by our QA on the dev server and passed testing. So I hope now we have some technical questions only to discuss.

demeritcowboy commented:
There's a couple things that need addressing but overall it looks good. Since it's a big PR there's lots of comments. Even for small PRs I'm not usually short on comments...

* General standards
    * [`r-explain`] __PASS__
    * [`r-doc`] __Issue__
      * Needs *two* docs PRs. One for the user guide and one for the dev docs for the hook. I don't think it needs to happen before merge.
        * One thing it could mention is that if you use the setting that prevents creating new contacts and it's the email sender that's new, then it uses the cron job user/contact, or whoever runs the job.
    * [`r-user`] __PASS__
      * The double-negatives in the setting label/description are ok because they were discussed in the lab ticket and it was agreed that would be ok. But see `r-doc` since in a docs PR there'd be more space to give examples.
    * [`r-run`] __PASS__
      * Ok. Yay! There is one thing about blank `subject`s which I've asked for clarification down in `r-code` below.
      * While I did test some of the new functionality, since you've already said it passes your own needs I only tested it lightly and my focus was mostly on confirming existing functionality and overall tech.
      * I also tested the existing emailProcessor hooks but not the new hook.
* Developer standards
    * [`r-tech`] __Issue__
      * I think maybe @eileenmcnaughton 's comments in the lab ticket about the deprecated build_mail_params might have been misinterpreted. I think she was making a suggestion about one place to do some code cleanup along the way, i.e. *completely removing it*. This PR just moves the deprecated block of code as-is to somewhere else and marks it not deprecated. That's probably a blocker to merging this. Given the scope size already at this point I would just put it back where it was for now.
      * The case subject additions weren't part of the lab ticket discussion at all. It's probably ok since the extra parts only kick in if you turn on the setting, just it's kind of a whole other *subject* on its own (pun intended).
        * As an aside: I can't believe I never noticed this before but `[0-9a-h]` has always been a typo I think. It's not possible to have a `g` or `h` in the hash. It doesn't hurt anything though.
    * [`r-code`] __Issue__
      * For schema changes the updated CRM/Core/DAO/MailSettings.php needs to committed. (There's supposed to be a check that tells you this but it's [not working](https://lab.civicrm.org/dev/core/-/issues/1408)).
      * In the upgrade script, to make upgrades safer and re-runnable, it should use the helper function addColumn() mentioned in the bullet points in the [dev docs upgrade section](https://docs.civicrm.org/dev/en/latest/framework/upgrade/#tips-prefer-simple-sql-semantics-over-apibaodao) and be a php script not a SQL script. Here is [an example](https://github.com/civicrm/civicrm-core/blob/58c72a10da01d86f790a62c54976cbf8726117d0/CRM/Upgrade/Incremental/php/FiveTwentyEight.php#L92)
        * At the current time this would go in the CRM/Upgrade/Incremental/php/FiveThirtyOne.php file, but as @mattwire noted it depends when this ultimately gets merged.
        * Similarly in the xml file the `<add>` is currently 5.31.
      * The `if` block at line 256 in EmailProcessor.php I'm not clear about. Why is subject included in this check? And then if it doesn't enter the `if` then it skips the setting of $matches which is used after the `if` so might have the wrong value. Before the patch $matches was always explicitly set to true or false in this spot.
      * The lines in the test that do assertTrue(empty) or assertTrue(!empty) can just be assertEmpty and assertNotEmpty.
        * (Just a comment) For line 213 since it's testing filing on cases then it would be a bonus if it could check the `case_id` member instead of `id` to confirm it filed on the case at the same time. But cases aren't enabled or set up at this point in the test, otherwise could just add `'return' => ['case_id']` to the api call and then check `assertNotEmpty($result['values'][0]['case_id'])`.
      * (Just a comment) The code style for set/get in CaseMail.php isn't wrong just isn't seen anywhere else in core that I know of. I can see it's a generic setter that silently does nothing if you give it a name that isn't already defined as a class member with a default value. I don't think there's any written standards about this type of thing.
    * [`r-maint`] __PASS__
      * Thanks for adding tests!
      * It would make more sense to me if they were in the EmailProcessorInboundTest.php file, but it's not a blocker. The EmailProcessorTest.php file seems to be for bounce processing.
    * [`r-test`] __PASS__

Lastly, regarding release date, and maybe this discussion belongs in chat or something: I can see an internal business reason for not wanting to manage patched sites, but a db schema change isn't a *technical* barrier to applying the patch and manually running some SQL. If you're worried about later doing a normal upgrade and it causing db problems, that would be handled by using the addColumn function in the upgrade script as mentioned above.

eileenmcnaughton commented:
Thanks for digging into this @demeritcowboy

ivan-compucorp commented:
Thanks @demeritcowboy for the review.

1. Needs two docs PRs.
- I think I need more details on that as I don't know how to create doc PRs.
1. I think maybe @eileenmcnaughton 's comments in the lab ticket about the deprecated build_mail_params might have been misinterpreted.
- Reverted the change. I think I don't have a big picture of the code and don't know how we can get rid of the function that is actually used.
1. It's not possible to have a g or h in the hash
- Fixed `CRM_Utils_Mail_CaseMail` class, now regexp is `[0-9a-f]`.
1. For schema changes the updated CRM/Core/DAO/MailSettings.php needs to committed.
- Afaik we should not do that, because this file would be regenerated before civicrm release, no?
1. Also technically sql/civicrm_generated.mysql needs to be updated, but should still work as-is just is out of sync.
- Afaik we should not do that, because this file would be regenerated before civicrm release, no? Anyway db upgrader should handle it I think.
1. In the upgrade script, to make upgrades safer and re-runnable, it should use the helper function addColumn()
- Moved db upgrader to php now and changed version to 5.31.0.
1. If it doesn't enter the if then it skips the setting of $matches which is used after the if so might have the wrong value. Before the patch $matches was always explicitly set to true or false in this spot.
- There is no point of creating empty activity which is not assigned to any contact, that's why this `if` was added. At the same time I've moved `$matches = FALSE;` before the `if` to avoid any ambiguity.
1. Line 19 in the .hlp file typo - should be "moved to the ignored folder".
- Fixed.
1. There's a couple places where it says "case token". Can we change that to "case hash" to avoid confusion with mail tokens?
- Fixed.
1. The lines in the test that do assertTrue(empty) or assertTrue(!empty) can just be assertEmpty and assertNotEmpty.
- Fixed.

I've also squashed commits, could you review again please?

demeritcowboy commented on CRM/Upgrade/Incremental/php/FiveThirtyOne.php:
```diff
@@ -74,6 +74,8 @@ public function upgrade_5_31_alpha1($rev) {
   public function upgrade_5_31_0($rev) {
     $this->addTask(ts('Upgrade DB to %1: SQL', [1 => $rev]), 'runSql', $rev);
+    $this->addTask('Add is_non_case_email_skipped column to civicrm_mail_settings', 'addColumn',
+      'civicrm_mail_settings', 'is_non_case_email_skipped', "TINYINT DEFAULT 0 NOT NULL COMMENT 'Skip emails which do not have a Case ID or Case hash'");
```
Yes it's hard to know which function it goes in - while it's still in master it would be 5_31_alpha1. When it gets branched to release-candidate it would be 5_31_beta1. The timing is difficult.

But the reason for the test fail is that one of the commits is trying to add the whole file instead of just changes. Maybe some git mixup.

eileenmcnaughton replied:
We wouldn't merge it to the rc once it's cut. If this makes 5.31 it is 5.31 - if not it's 5.32 - the core release cycle is regular enough that missing a release shouldn't be too big a delay - but if you want to deploy to clients before it is in an official release you can patch their release / add the columns via sql on their release.

ivan-compucorp replied:
So we can leave the code as is for now, right?

demeritcowboy replied:
As of right this minute, it should be inside the existing 5_31_alpha1 function. But maybe wait and see if anyone else has comments on the PR? For testing I moved them there and they ran properly.

demeritcowboy commented on CRM/Utils/Mail/EmailProcessor.php:
```diff
@@ -241,16 +251,20 @@ public static function _process($civiMail, $dao, $is_create_activities) {
-            echo "Processed as Activity: {$mail->subject}\n";
+          // Create activity if its not empty.
+          $matches = FALSE;
+          if (!empty($params['subject']) || !empty($params['target_contact_id'])) {
```
I think I'm still not clear on this part. Suppose you turn on the setting to skip creating contacts. So $params['target_contact_id'] will be empty. Suppose also that the `from` address is already a contact in civi. And suppose also that it's an email with no subject but it has a body. It should file that email, right?

ivan-compucorp replied:
Okay will revert this for now.

demeritcowboy replied:
Thanks

eileenmcnaughton commented:
The jenkins errors are due to it failing to apply - probably a rebase & force push will address

  Applying: CATL-1614: Add new field to the mail account form - Do not create new contacts when filing emails             
  Patch failed at 0001 CATL-1614: Add new field to the mail account form - Do not create new contacts when filing emails  
  When you have resolved this problem, run "git am --continue".                                                           
  If you prefer to skip this patch, run "git am --skip" instead.                                                          
  To restore the original branch and stop patching, run "git am --abort".                                                 
  [[ STDERR ]]                                                                                                            
  error: CRM/Upgrade/Incremental/php/FiveThirtyOne.php: already exists in index                                           
  hint: Use 'git am --show-current-patch' to see the failed patch

ivan-compucorp commented:
@demeritcowboy I've fixed jenkins error, mentioned by Eileen, is everything okay now?

demeritcowboy commented:
I don't have anything more on the content of the PR to add other than some tidy-up notes below.

1. For the DAO CRM/Core/DAO/MailSettings.php and civicrm_generated.mysql, they do need to be committed, but it could also be a separate PR after.

2. For the docs PRs:
  a. For the hook, you would make a PR against the https://lab.civicrm.org/documentation/docs/dev/ repo, adding a file in here: https://lab.civicrm.org/documentation/docs/dev/-/tree/master/docs/hooks, and then also updating `list.md` in there. If you're not familiar with gitlab it's the same as github - make a fork, make a branch, commit to your fork/branch, then just note they call it merge requests instead of pull requests.
  b. For the user/admin, it would be against the sysadmin guide. This looks like a good spot https://lab.civicrm.org/documentation/docs/sysadmin/-/tree/master/docs/setup/civimail#adding-an-incoming-email-account-for-processing-bounces-andor-email-to-activities

3. The current failing test (deleteByCacheKey) is a known flaky test. Can ignore it.

seamuslee001 commented:
Merging based on @demeritcowboy 's review

eileenmcnaughton commented:
@ivan-compucorp we have merged this in advance of the docs to hit the 5.31 rc - but please follow up by doing the docs...

ivan-compucorp commented:
Thanks @demeritcowboy, @eileenmcnaughton :tada: 
Will add documentation soon.

ivan-compucorp commented:
Added documentation:
1. https://lab.civicrm.org/documentation/docs/dev/-/merge_requests/859
1. https://lab.civicrm.org/documentation/docs/sysadmin/-/merge_requests/272

Is there anything else that should be done @demeritcowboy?

demeritcowboy commented:
Nothing else - the elves came by at night and updated some things: https://github.com/civicrm/civicrm-core/pull/18711
